End-to-End Streaming Tests in botbuilder-js

Zerryth opened the pull request:
Fixes #2226

## Description
PR automates the testing of botframework-streaming's compatibility with browsers. 

- This is done using a React app that uses DLJS and a bot that both point to the latest preview versions of botframework-streaming. They communicate by enabling DL ASE. 
- Tests are automated using Azure Pipelines as well as leveraging Selenium WebDriver to automate browser testing

## Specific Changes
*E2E Streaming Tests Architecture*
![image](https://user-images.githubusercontent.com/35248895/88363581-6209ed80-cd35-11ea-8312-3c7a5f3f5303.png)


[E2E Streaming Tests Pipeline here](https://fuselabs.visualstudio.com/SDK_v4/_build?definitionId=1018&_a=summary) (currently under temporary name "[experimental]E2E_Streaming_Test_Ashley")

Running the pipeline will:
- **Provision & Deploy Projects**:
    - **Bot**: point bot to use the daily build of botframework-streaming (latest preview version)
    - **React App**:
        - Provision a brand new react app via `npx create-react-app` command
        - Set up the React app to use WebChat (`ReactWebChat`)
        - Install a custom version of DirectLineJS (DLJS) into the app
            - Configures DLJS to use the latest botframework-streaming preview version
- **Run Tests**
    - ~Downloads the browser-appropriate Driver necessary for Selenium WebDriver to automate testing in a browser (Chrome)~ Use the windows-2019 image's pre-installed of WebDriver & ChromeDriver to automate testing in browser
    - Runs the tests that have React and bot talk to each other

See [README](https://github.com/microsoft/botbuilder-js/blob/Zerryth/streaming-e2e-tests/libraries/streaming-e2e-tests/README.md) for further details of PR, if wanted.

## Testing

<!-- If you are adding a new feature to a library, you must include tests for your new code. -->
It's all green!
![image](https://user-images.githubusercontent.com/35248895/88363782-02f8a880-cd36-11ea-943c-3ef861c3bc85.png)

___

This PR is already large enough, and it only tests 1 scenario (webpack'd project + chrome browser). In future expansions of streaming tests, we could include the following (detailed per William):

*Different Build Scenarios*:
1. Use `create-react-app` (CRA) without WebChat, just talk to bot using DLJS
2. Use webpack without CRA
3. Add Rollup in addition to webpack
 
*Different Browsers*:
- Edge UWP
- Edge Chromium
- Firefox
- IE11
- macOS Safari
- iOS Safari
- Android Chrome

The expansion should go more quickly now that this PR lays down the foundational infrastructure for E2E testing of streaming bits (at least for adding additional browsers to test against).

compulim commented on libraries/streaming-e2e-tests/bot/utils/generateDirectLineToken.js:
```diff
@@ -0,0 +1,47 @@
+      trustedOrigins: [
+        'https://compulim.github.io/',
+        'https://microsoft.github.io/',
+        'https://webchat-mockbot2.azurewebsites.net/',
```
Clean up the `trustedOrigins`. 😉

Zerryth replied:
cleared!

compulim commented on libraries/streaming-e2e-tests/bot/.env:
```diff
@@ -0,0 +1,4 @@
+MicrosoftAppId=
+MicrosoftAppPassword=
+DIRECT_LINE_SECRET=
```
Add `DIRECT_LINE_URL` here?

Zerryth replied:
n/a now that we are only going to request token from the DL ASE token gen endpoint

compulim commented on libraries/streaming-e2e-tests/bot/utils/generateDirectLineToken.js:
```diff
@@ -0,0 +1,47 @@
+
+module.exports = async function (
+  directLineSecret = process.env.DIRECT_LINE_SECRET,
+  { domain = process.env.DIRECT_LINE_URL || 'https://directline.botframework.com/', userId = createUserId() } = {}
```
Should this function always generate a DL ASE token?

```js
domain = `https://${process.env.WEBSITE_HOSTNAME}.azurewebsites.net/.bot/`
```

Zerryth replied:
Hmm yeah. Considering we want to test DL ASE here 😅 
The copy pasta strikes again!

Changing it to the DL ASE token gen endpoint. Thanks William

compulim commented on libraries/streaming-e2e-tests/bot/utils/generateDirectLineToken.js:
```diff
@@ -0,0 +1,47 @@
+        'https://microsoft.github.io/',
+        'https://webchat-mockbot2.azurewebsites.net/',
+        'http://localhost',
+        `https://${WEBSITE_HOSTNAME}.azurewebsites.net/`
```
```suggestion
        `https://${process.env.WEBSITE_HOSTNAME}.azurewebsites.net/`
```

(So we can get rid of one-use variable)

Zerryth replied:
Updated!

Zerryth replied:
cleaned up trusted origins

compulim commented on libraries/streaming-e2e-tests/bot/utils/renewDirectLineToken.js:
```diff
@@ -0,0 +1,27 @@
+module.exports = async function (
+    token,
+    { domain = process.env.DIRECT_LINE_URL || 'https://directline.botframework.com/' } = {}
```
Same as above? Should we always renew DL ASE type of token?

Zerryth replied:
ditto, no longer applicable (using DL ASE token gen endpoint, as your other comment pointed out)

compulim commented on libraries/streaming-e2e-tests/bot/web.config:
```diff
@@ -0,0 +1,60 @@
+    <rewrite>
+      <rules>
+        <!-- Do not interfere with Direct Line App Service Extension -->
+        <rule name ="DLASE" stopProcessing="true">
```
```suggestion
        <rule name="DLASE" stopProcessing="true">
```

compulim commented on libraries/streaming-e2e-tests/bot/web.config:
```diff
@@ -0,0 +1,60 @@
+          </conditions>
+        </rule>
+        <!-- Do not interfere with co-hosted Web Chat -->
+        <rule name ="WEBCHAT" stopProcessing="true">
```
Clean up.

Zerryth replied:
Removed `WEBCHAT` rule ✔️

compulim commented on libraries/streaming-e2e-tests/build/create-react-app.yml:
```diff
@@ -0,0 +1,37 @@
+      Get-Content -Path "./.env"
+    displayName: "Write bot's website host name to React app's .env"
+    workingDirectory: $(RootReactPiecesDir)/react-app
+
```
Run `npm run build` here.

Then pack the `/build/**/*` and publish it as artifact.

compulim commented on libraries/streaming-e2e-tests/build/deploy-react-app.yml:
```diff
@@ -0,0 +1,19 @@
+steps:
+  - script: npm run build
```
Put the `npm run build` in the previous YAML.

compulim commented on libraries/streaming-e2e-tests/build/run-tests.yml:
```diff
@@ -0,0 +1,15 @@
+jobs:
+  - job: Install_Drivers_And_Run_Tests
+    variables:
+      ChromeDriverDownloadUrl: 'https://chromedriver.storage.googleapis.com/84.0.4147.30/chromedriver_win32.zip'
```
Think about how to sync the version with Microsoft-hosted agent.

https://github.com/actions/virtual-environments/blob/master/images/win/Windows2019-Readme.md

compulim commented on libraries/streaming-e2e-tests/build/create-react-app.yml:
```diff
@@ -0,0 +1,37 @@
+    displayName: 'create-react-app'
+    workingDirectory: $(RootReactPiecesDir)
+
+  - powershell: npm install botframework-webchat
```
```suggestion
  - powershell: npm install botframework-webchat@4.9.2
```

Please lock down on version of Web Chat. This is because the DOM tree of Web Chat may change from time to time and it will affect your test code.

(Or file an issue to ask Web Chat to publish a page object library)

compulim commented on libraries/streaming-e2e-tests/tests/streaming-e2e.test.js:
```diff
@@ -0,0 +1,102 @@
+
+async function echoMessageInBrowser(driver) {
+  try {
+    await driver.get(process.env.REACT_APP_ENDPOINT);
```
May be you can consider hosting the server locally, so we don't need to ZipDeploy to Azure Web App, and then hosting the server publicly.

Zerryth replied:
Agreed!

Per PR review meeting yesterday, originally published React App, as Steven and I weren't sure we would get automated browser testing in on time for R10, so we deployed React App to at least be able to publish manually

Created [issue#2607](https://github.com/microsoft/botbuilder-js/issues/2607) to implement locally-hosted React App for this pipeline for R11 (not enough time for R10)

Zerryth replied:
Whoops, now that's a dupe. Steven beat me to it!

Here's [the actual issue then](https://github.com/microsoft/botbuilder-js/issues/2603), and I'll close the duplicate 😆

Zerryth commented:
Updated:
- Removed chromedriver installation ✔️ 
    - [The image already has chrome & chromedriver installed ](https://github.com/microsoft/azure-pipelines-image-generation/issues/861#issuecomment-490601694l)and maintains compatible versions for them
    - w00t! taken care of before R11 🎊 

- Removed `OnMembersAdded` handler from bot to speed up testing time dramatically in Azure Pipeline environment
   <details>
         <summary>Details</summary>

  - Times running Selenium tests vary *wildly* on in Pipelines (27s - 3+ min) versus on my own computer (20s)--we're at the mercy of whatever build machines the pipeline's tasks land on and pray for a good one
    - Removing extra cruft so that tests could run faster and complete in a more reasonable time
   </details>

- Cleaned up `configure-to-bbjs-daily.yml` (it repeats a step one less time) ✔️ 
    - To work around the issue of Azure Pipelines not expanding nested YAML variables, just did the get&set from w/in one PS script, so we didn't need to repeat steps in YAML

- Mocha tests now use pipeline variable instead of me writing value to `.env` ✔️ 

___
Could not get: react app to pull from pipeline variables ❌ 

- I imagine this has something to do with the react app being in browser and/or not having access to the pipeline variables once deployed to App Service
- For R11, once we do the work for hosting the react app locally, I'll see if it's possible then...but it seems like from researching, their docs [pointed to using a `.env` for react apps](https://create-react-app.dev/docs/adding-custom-environment-variables/#adding-development-environment-variables-in-env) and *must* be prefixed with `REACT_APP_` in the variable name or using webpack with DefinePlugin

Zerryth commented:
Hmm, looks like I need to provide more timeout time, so I'll update that too. Like I mentioned, sometimes it completes within 30 s., sometimes it takes minutes, just depending on the agent it lands on...
